File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/globalcontexts/GlobalContextsConfigurationInterface.java
package com.snowplowanalytics.snowplow.internal.globalcontexts;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.snowplowanalytics.snowplow.globalcontexts.GlobalContext;

import java.util.Set;

public interface GlobalContextsConfigurationInterface {

    /**
     * @return Set of tags associated to added GlobalContexts.
     */
    @NonNull
    Set<String> getTags();

    /**
     * Add a GlobalContext generator to the configuration of the tracker.
     * @param tag The label identifying the generator in the tracker.
     * @param contextGenerator The GlobalContext generator.
     * @return Whether the adding operation has succeeded.
     */
    boolean add(@NonNull String tag, @NonNull GlobalContext contextGenerator);

    /**
     * Remove a GlobalContext generator from the configuration of the tracker.
     * @param tag The label identifying the generator in the tracker.
     * @return Whether the removing operation has succeded.
     */
    @Nullable
    GlobalContext remove(@NonNull String tag);
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/core/globalcontexts/GlobalContextsConfigurationInterface.kt
/*
 * Copyright (c) 2015-2023 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.core.globalcontexts

import com.snowplowanalytics.snowplow.globalcontexts.GlobalContext

interface GlobalContextsConfigurationInterface {
    /**
     * @return Set of tags associated to added GlobalContexts.
     */
    val tags: Set<String?>

    /**
     * Add a GlobalContext generator to the configuration of the tracker.
     * @param tag The label identifying the generator in the tracker.
     * @param contextGenerator The GlobalContext generator.
     * @return Whether the adding operation has succeeded.
     */
    fun add(tag: String, contextGenerator: GlobalContext): Boolean

    /**
     * Remove a GlobalContext generator from the configuration of the tracker.
     * @param tag The label identifying the generator in the tracker.
     * @return Whether the removing operation has succeeded.
     */
    fun remove(tag: String): GlobalContext?
}
